BlueprintUI/Sources/Layout/ConstrainedAspectRatio.swift

```diff
@@ -2,35 +2,113 @@ import UIKit
+        /// dimension, the element will fill the constrained dimension.
+        case fillParent
 
-        func constrain(size: CGSize, to aspectRatio: AspectRatio) -> CGSize {
-            let constrainedHeight = aspectRatio.height(forWidth: size.width)
-            let constrainedWidth = aspectRatio.width(forHeight: size.height)
+        /// The content will be sized to fit within its parent while maintaining the specified
+        /// aspect ratio.
+        ///
+        /// If the parent is unconstrained in all dimensions, the content size will be used for
+        /// measurement and will behave like `fitContent`. If the parent is unconstrained in one
+        /// dimension, the element will fit the constrained dimension.
+        case fitParent
+
+        /// The content will grow in whichever dimension is needed to maintain the aspect ratio,
+        /// while ensuring the content still fits.
+        ///
+        /// This mode does not take the parents applied size constraint into account, so the parent
+        /// may ultimately layout the element without the constrained aspect ratio.
+        case fitContent
+
+        /// The content will shrink in whichever dimension is needed to maintain the aspect ratio,
+        /// meaning it may be smaller in one dimension than the measured size.
+        ///
+        /// This mode does not take the parents applied size constraint into account, so the parent
+        /// may ultimately layout the element without the constrained aspect ratio.
+        case shrinkContent
+
+        func constrain(
+            contentSize: CGSize,
+            in constraint: SizeConstraint,
+            to aspectRatio: AspectRatio
+        ) -> CGSize {
+            let availableHeight: CGFloat
+            let availableWidth: CGFloat
 
             switch self {
-            case .fill:
-                if constrainedWidth > size.width {
-                    return CGSize(width: constrainedWidth, height: size.height)
-                } else if constrainedHeight > size.height {
-                    return CGSize(width: size.width, height: constrainedHeight)
+            case .fillParent,
+                 .fitParent:
+                availableHeight = constraint.height.constrainedValue ?? contentSize.height
+                availableWidth = constraint.width.constrainedValue ?? contentSize.width
+            case .fitContent,
+                 .shrinkContent:
+                availableHeight = contentSize.height
+                availableWidth = contentSize.width
+            }
+
+            let constrainedHeight = aspectRatio.height(forWidth: availableWidth)
+            let constrainedWidth = aspectRatio.width(forHeight: availableHeight)
+
+            switch self {
+            case .fillParent:
+                if constraint.width.constrainedValue == nil &&
+                    constraint.height.constrainedValue == nil {
```

Comment:
Should this be an `&&`, or an `||` in terms of the fallback? Though I guess if it was an `||`, unconstrained parents like scroll views wouldn't work.

Reply:
The cases for a single dimension are handled right below this statement, so it should be `&&`